src/rtpa-server.1
.\" ##########################################################################
.\" ####                                                                  ####
.\" ####                      RTP Audio Server Project                    ####
.\" ####                    ============================                  ####
.\" ####                                                                  ####
.\" #### RTP Audio Server Manpage                                         ####
.\" ####                                                                  ####
.\" ####           Copyright (C) 1999-2026 by Thomas Dreibholz            ####
.\" ####                                                                  ####
.\" #### Contact:                                                         ####
.\" ####    EMail: thomas.dreibholz@gmail.com                             ####
.\" ####    WWW:   https://www.nntb.no/~dreibh/rtpaudio                   ####
.\" ####                                                                  ####
.\" #### ---------------------------------------------------------------- ####
.\" ####                                                                  ####
.\" #### This program is free software: you can redistribute it and/or    ####
.\" #### modify it under the terms of the GNU General Public License as   ####
.\" #### published by the Free Software Foundation, either version 3 of   ####
.\" #### the License, or (at your option) any later version.              ####
.\" ####                                                                  ####
.\" #### This program is distributed in the hope that it will be useful,  ####
.\" #### but WITHOUT ANY WARRANTY; without even the implied warranty of   ####
.\" #### MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    ####
.\" #### GNU General Public License for more details.                     ####
.\" ####                                                                  ####
.\" #### You should have received a copy of the GNU General Public        ####
.\" #### License along with this program.  If not, see                    ####
.\" #### <http://www.gnu.org/licenses/>.                                  ####
.\" ####                                                                  ####
.\" ##########################################################################
.\"
.\" ###### Setup ############################################################
.Dd November 29, 2024
.Dt rtpa-server 1
.Os rtpa-server
.\" ###### Name #############################################################
.Sh NAME
.Nm rtpa-server
.Nd RTP Audio Server
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm rtpa-server
.Op Fl port=port
.Op Fl directory=path
.Op Fl timeout=secs
.Op Fl maxpktsize=bytes
.Op Fl disable-qm
.Op Fl enable-qm
.Op Fl disable-ls
.Op Fl enable-ls
.Op Fl force-ipv4
.Op Fl use-ipv6
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm rtpa-server
is the RTP Audio server application.
.Pp
.\" ###### Arguments ########################################################
.Sh OPTIONS
The following options may be provided:
.Bl -tag -width indent
.It Fl port=port
TBD.
.El
.\" ###### Arguments ########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It rtpa-server
.It rtpa-server -sctp
.It rtpa-server -directory=/path/to/media/directory
.El
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz
.br
https://www.nntb.no/~dreibh/rtpaudio
.br
mailto://thomas.dreibholz@gmail.com
.br


ChangeLog
commit 6f741a1851b4ce35a2dc7dbf7f20aac2b716111c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 14:17:00 2023 +0100

    Updated packaging configuration.

commit 01ff4b24a92dc6518e8bdda5621420c0e18b33eb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 14:07:46 2023 +0100

    FreeBSD build fix.

commit ce4414517294195ef6578d8b66ef76babebe92ef
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 13:47:27 2023 +0100

    Packaging update.

commit 65eaa6e46d8b547a6536aeb0adb50446842eb8c5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 12:32:29 2023 +0100

    Improved FreeBSD test script.

commit 935cee09e85c735aba38035ad3935824be1c0f4d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 10:29:54 2023 +0100

    Updated LSM entry.

commit f4f6bb87107bd8bf6c179c28e1dae1e50eb37e2e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 10:20:27 2023 +0100

    Updated packaging configuration.

commit dac93f6bb5bd8923b4156ea87ea89890f9363630
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 17:18:26 2023 +0100

    Updated LSM entry.

commit 0fca1af2d6163cefedf9c5e74cd3241a06492720
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 16:54:41 2023 +0100

    New release rtpaudio-2.0.7.

commit 09a51260dfbf72af1f0fea2962e867e047f17e69
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 13:59:49 2023 +0100

    Added lintian override.

commit 037dbc4ead53e77329f1740b744173c387cf9855
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 13:49:44 2023 +0100

    FreeBSD build fix.

commit f8f3d56922297b3b3bd50fc40760593930bb012a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 11:42:31 2023 +0100

    Dependency improvements.

commit 4bf74c50c0891e39d54cd5705852bf76bc4b34fa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 10:53:51 2023 +0100

    Packaging clean-ups.

commit a55536dd6f962e04d15ecfd0e19f610311d533b7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 09:52:23 2023 +0100

    Updated build scripts

commit 24b6acf828c23e20549cd63a45401a08fbf55936
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 09:33:15 2023 +0100

    Updated build scripts

commit fa13ea861b4ea0f939cd26f55f95bd6855b53ec1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 17:50:51 2023 +0100

    Workflows update.

commit 801227ed26423d950eb4ee3fbb9f6a83dbb08a84
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:27:16 2023 +0100

    Minor manpage updates.

commit 22e407d96e0d93c1ff1988004b59c62f03a02abd
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:22:36 2023 +0100

    Header updates.

commit 59469efa6aa8aaa3399f97042936b41fa07d3775
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:18:09 2023 +0100

    Updated CI scripts

commit ec0a4ecaa4aefa6f2f0e5827affa44626b0766a2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:18:08 2023 +0100

    Updated build scripts

commit 5bec64693ac502fe6f8c4ac42e9d7cacfbf148c8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 13:23:07 2023 +0100

    URL change from UDE to NNTB.

commit c4cfbd10269fb83ba954693d61a1bd62d97db79b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 12:13:30 2023 +0100

    Updated autogen.sh.

commit 81cebd8e6549ed09947ac103379a16ed9385f6da
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:55:31 2023 +0100

    Updated build scripts

commit bcd562c874e2730fec7363db09545076a630558f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:52:40 2023 +0100

    Updated CI scripts

commit 4776b568eeb87dd91be1216940918536da0c053d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:52:34 2023 +0100

    Travis CI is dead. Removed obsolete Travis CI scripts and configuration.

commit b2328c6151bb9605cdeae292136ab1d7bda13940
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Sep 11 13:47:02 2022 +0200

    New release rtpaudio-2.0.6.

commit ce043eb413eae3a35979e495e421b1ed2f778b43
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:47:32 2022 +0200

    Removed Vcs-* fields from debian/control.

commit c02b46d03953abd5b0e1017e1bb5ef9ae0f54762
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:21:06 2022 +0200

    RPM fix.

commit 3e20567e717c452ff28e0731dbdaf084516322c6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:19:33 2022 +0200

    Updated CI scripts

commit 641a5f665964ef8bdb0b8c662c2de185f565822a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:19:32 2022 +0200

    Updated build scripts

commit 4d4c7a4591572142414b8df7af5194abb8b1c06f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 5 11:41:26 2022 +0200

    Fix for RPM spec: rtpaudio-all is build now.

commit 43f37568b3ea9828919bcd070a07c7a6db14fec6
Author: Thomas Dreibholz <dreibh@simula.no>

src/rtpa-server.cc
// ##########################################################################
// ####                                                                  ####
// ####                      RTP Audio Server Project                    ####
// ####                    ============================                  ####
// ####                                                                  ####
// #### Audio Server                                                     ####
// ####                                                                  ####
// ####           Copyright (C) 1999-2026 by Thomas Dreibholz            ####
// ####                                                                  ####
// #### Contact:                                                         ####
// ####    EMail: thomas.dreibholz@gmail.com                             ####
// ####    WWW:   https://www.nntb.no/~dreibh/rtpaudio                   ####
// ####                                                                  ####
// #### ---------------------------------------------------------------- ####
// ####                                                                  ####
// #### This program is free software: you can redistribute it and/or    ####
// #### modify it under the terms of the GNU General Public License as   ####
// #### published by the Free Software Foundation, either version 3 of   ####
// #### the License, or (at your option) any later version.              ####
// ####                                                                  ####
// #### This program is distributed in the hope that it will be useful,  ####
// #### but WITHOUT ANY WARRANTY; without even the implied warranty of   ####
// #### MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    ####
// #### GNU General Public License for more details.                     ####
// ####                                                                  ####
// #### You should have received a copy of the GNU General Public        ####
// #### License along with this program.  If not, see                    ####
// #### <http://www.gnu.org/licenses/>.                                  ####
// ####                                                                  ####
// ##########################################################################


#include "tdsystem.h"
#include "tdsocket.h"
#include "rtpsender.h"
#include "rtcppacket.h"
#include "rtcpreceiver.h"
#include "rtcpabstractserver.h"
#include "audioclientapppacket.h"
#include "audioserver.h"
#include "tools.h"
#include "breakdetector.h"

#define WITH_QOSMGR
#ifdef WITH_QOSMGR
#include "bandwidthmanager.h"
#include "servicelevelagreement.h"
#if 0
#include "roundtriptimepinger.h"
#endif
#endif

#include <fstream>



// Fast Break: Disable break detector to debug thread deadlocks
// #define FAST_BREAK


static Socket*                rtcpServerSocket  = NULL;
static RTCPReceiver*          rtcpReceiver      = NULL;
static AudioServer*           server            = NULL;
static BandwidthManager*      qosManager        = NULL;
static ServiceLevelAgreement* sla               = NULL;
static Socket*                pingSocket4       = NULL;
static Socket*                pingSocket6       = NULL;
static RoundTripTimePinger*   pinger            = NULL;
static std::ofstream*         logStream         = NULL;


void cleanUp(const cardinal exitCode = 0);


// ###### Initialize ########################################################
void initAll(const char*    directory,
             const card16   port,
             const card64   timeout,
             const cardinal maxPacketSize,
             const bool     lossScalability,
             const bool     useSCTP)
{
   const InternetAddress localAddress(port);
   rtcpServerSocket = new Socket(Socket::IP,
                                 useSCTP ? Socket::SeqPacket : Socket::Datagram,
                                 useSCTP ? Socket::SCTP : Socket::Default);
   if(rtcpServerSocket == NULL) {
      std::cerr << "ERROR: Server::initAll() - Out of memory!" << std::endl;
      cleanUp(1);
   }
   if(rtcpServerSocket->bind(localAddress) == false) {
      std::cerr << "ERROR: Server::initAll() - Unable to bind socket!" << std::endl;
      cleanUp(1);
   }
   if(useSCTP) {
      rtcpServerSocket->listen(10);
   }
   server = new AudioServer(qosManager, maxPacketSize, useSCTP);
   if(server == NULL) {
      std::cerr << "ERROR: Server::initAll() - Out of memory!" << std::endl;
      cleanUp(1);
   }
   server->setDefaultTimeout(timeout);
   server->setLossScalability(lossScalability);
   rtcpReceiver = new RTCPReceiver(server,rtcpServerSocket);
   if(rtcpReceiver == NULL) {
      std::cerr << "ERROR: Server::initAll() - Out of memory!" << std::endl;
      cleanUp(1);
   }
   if(server->start() == false) {
      std::cerr << "ERROR: Server::initAll() - Unable to start server thread!" << std::endl;
      cleanUp(1);
   }
   if(rtcpReceiver->start() == false) {
      std::cerr << "ERROR: Server::initAll() - Unable to start RTCP receiver thread!" << std::endl;
      cleanUp(1);
   }

   // ====== Change directory ===============================================
   if(directory != NULL) {
      if(chdir(directory) != 0) {
         std::cerr << "ERROR: Server::initAll() - Unable to change directory!" << std::endl;
         cleanUp(1);
      }
   }
   else {
      directory = "./";
   }
}


// ###### Clean up ##########################################################
void cleanUp(const cardinal exitCode)
{
   if(rtcpReceiver != NULL) {
      rtcpReceiver->stop();
      delete rtcpReceiver;
      rtcpReceiver = NULL;
   }
   if(server != NULL) {
      server->stop();
      delete server;
      server = NULL;
   }
   if(rtcpServerSocket != NULL) {
      delete rtcpServerSocket;
      rtcpServerSocket = NULL;
   }
   if(qosManager != NULL) {
      delete qosManager;
      qosManager = NULL;
   }
   if(exitCode == 0) {
      std::cout << "Terminated!" << std::endl;
   }
   exit(exitCode);
}


// ###### Main program ######################################################
int main(int argc, char* argv[])
{
   // ===== Initialize ======================================================
   bool     optForceIPv4           = false;
   bool     optUseSCTP             = false;
   bool     lossScalability        = true;
   bool     disableQM              = false;
   double   fairnessSession        = 0.0;
   double   fairnessStream         = 1.0;
   bool     prEnabled              = true;
   double   prReservedPortion      = 0.1;
   double   prUtilizationTolerance = 0.05;
   double   prMaxRemappingInterval = 5000000.0;
   cardinal maxRUPoints            = 32;
   double   utThreshold            = 0.01;
   card64   bwThreshold            = (card64)-1;
   double   sdTolerance            = 50000.0;
   bool     unlayered              = false;
   cardinal maxPacketSize          = 1500;
   card64   timeout                = 10000000;
   card16   port                   = RTPAudioDefaultPort;
   char*    logName                = NULL;
   String   slaFile("SLA.config");
   String   directory;


   // ====== Read configuration from file ===================================
   FILE* inputFD = fopen("AudioServer.config","r");
   if(inputFD != NULL) {
      char str[256];
      char* result = fgets((char*)&str,256,inputFD);
      cardinal line = 0;
      while( (result != NULL) && (!feof(inputFD)) ) {
         line++;
         const cardinal inputLength = strlen((char*)&str);
         if(inputLength > 1) {
            str[inputLength - 1] = 0x00;
            switch(str[0]) {
               // ====== Line is a comment =====================================
               case '#':
                break;

               // ====== Line is a setting =====================================
               default:
                  const String input(str);
                  String name;
                  String value;
                  if(input.scanSetting(name,value)) {
                     if(name == "PORT") {
                        int portNumber;
                        if(sscanf(value.getData(),"%d",&portNumber) != 1) {
                           std::cerr << "ERROR: Bad port setting, "
                                   "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Port = <number>" << std::endl;
                           exit(1);
                        }
                        port = (card16)portNumber;
                     }
                     else if(name == "TIMEOUT") {
                        int timeoutValue;
                        if(sscanf(value.getData(),"%d",&timeoutValue) != 1) {
                           std::cerr << "ERROR: Bad timeout setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Timeout = <seconds>" << std::endl;
                           exit(1);
                        }
                        timeout = 1000000 * (card64)timeoutValue;
                     }
                     else if(name == "MAX PACKET SIZE") {
                        int size;
                        if(sscanf(value.getData(),"%d",&size) != 1) {
                           std::cerr << "ERROR: Bad maximum packet size setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Max Packet Size = <Bytes>" << std::endl;
                           exit(1);
                        }
                        maxPacketSize = (cardinal)size;
                     }
                     else if(name == "DIRECTORY") {
                        directory = value;
                     }
                     else if(name == "DISABLE QOS MANAGER") {
                        int off;
                        if(sscanf(value.getData(),"%d",&off) != 1) {
                           std::cerr << "ERROR: Bad QoS manager setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Disable QoS Manager = <0|1>" << std::endl;
                           exit(1);
                        }
                        disableQM = (off != 0) ? true : false;
                     }
                     else if(name == "QOS MANAGER") {
                        int on;
                        if(sscanf(value.getData(),"%d",&on) != 1) {
                           std::cerr << "ERROR: Bad QoS manager setting, "
                                   "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: QoS Manager = <0|1>" << std::endl;
                           exit(1);
                        }
                        disableQM = (on != 0) ? false : true;
                     }
                     else if(name == "LOSS SCALABILITY") {
                        int on;
                        if(sscanf(value.getData(),"%d",&on) != 1) {
                           std::cerr << "ERROR: Bad loss scalability setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Loss Scalability = <0|1>" << std::endl;
                           exit(1);
                        }
                        lossScalability = (on != 0) ? true : false;
                     }
                     else if(name == "FORCE IPV4") {
                        int on;
                        if(sscanf(value.getData(),"%d",&on) != 1) {
                           std::cerr << "ERROR: Bad force IPv4 setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: Force IPv4 = <0|1>" << std::endl;
                           exit(1);
                        }
                        optForceIPv4 = (on != 0) ? true : false;
                     }
                     else if(name == "SET PARTIAL REMAPPING") {
                        int    on;
                        if(sscanf(value.getData(),
                                  "%d R=%lf T=%lf X=%lf",
                                  &on,&prReservedPortion,&prUtilizationTolerance,
                                  &prMaxRemappingInterval) > 0) {
                           prEnabled = (on != 0);
                        }
                     }
                     else if(name == "SET FAIRNESS") {
                        sscanf(value.getData(),
                               "%lf %lf",
                               &fairnessSession,&fairnessStream);
                     }
                     else if(name == "SET QOS OPTIMIZATION") {
                        int maxInt;
                        if(sscanf(value.getData(),
                                  "%u U=%lf B=%llu T=%lf",
                                  &maxInt,&utThreshold,
                                  (unsigned long long*)&bwThreshold,
                                  &sdTolerance) > 0) {
                           maxRUPoints = (cardinal)maxInt;
                        }
                     }
                     else if(name == "SCTP") {
                        int on;
                        if(sscanf(value.getData(),"%d",&on) != 1) {
                           std::cerr << "ERROR: Bad SCTP setting, "
                                        "line " << line << "!" << std::endl;
                           std::cerr << "       Syntax: SCTP = <0|1>" << std::endl;
                           exit(1);
                        }
                        optUseSCTP = (on != 0) ? true : false;
                     }
                     else {
                        std::cerr << "ERROR: Unknown option <"
                                  << name << " = " << value << ">, "
                                  "line " << line << "!" << std::endl;
                        exit(1);
                     }
                  }
                break;
            }
         }
         result = fgets((char*)&str,256,inputFD);
      }
      fclose(inputFD);
   }


   // ===== Check arguments =================================================
   for(cardinal i = 1;i < (cardinal)argc;i++) {
      if(!(strcasecmp(argv[i],"-force-ipv4")))           optForceIPv4 = true;
      else if(!(strcasecmp(argv[i],"-use-ipv6")))        optForceIPv4 = false;
      else if(!(strcasecmp(argv[i],"-sctp")))            optUseSCTP   = true;
      else if(!(strcasecmp(argv[i],"-nosctp")))          optUseSCTP   = false;
      else if(!(strncasecmp(argv[i],"-port=",6)))        port      = (card16)atol(&argv[i][6]);
      else if(!(strncasecmp(argv[i],"-timeout=",9)))     timeout   = 1000000 * (card64)atol(&argv[i][9]);
      else if(!(strncasecmp(argv[i],"-maxpktsize=",12))) maxPacketSize = (cardinal)atol(&argv[i][12]);
      else if(!(strcasecmp(argv[i],"-disable-qm")))      disableQM = true;
      else if(!(strcasecmp(argv[i],"-enable-qm")))       disableQM = false;
      else if(!(strcasecmp(argv[i],"-disable-ls")))      lossScalability = false;
      else if(!(strcasecmp(argv[i],"-enable-ls")))       lossScalability = true;
      else if(!(strncasecmp(argv[i],"-sla=",5)))         slaFile       = &argv[i][5];
      else if(!(strncasecmp(argv[i],"-log=",5)))         logName      = &argv[i][5];
      else if(!(strncasecmp(argv[i],"-directory=",11)))  directory = String(&argv[i][11]);
      else {
         std::cerr << "Usage: " << argv[0] << " {-port=port} {-directory=path} {-manager=host:port} {-timeout=secs} {-maxpktsize=bytes} {-disable-qm|-enable-qm} {-disable-ls|-enable-ls} {-force-ipv4|-use-ipv6}" << std::endl;
         exit(1);
      }
   }
   if(optForceIPv4) {
      if(InternetAddress::UseIPv6 == true) {
         InternetAddress::UseIPv6 = false;
         std::cerr << "NOTE: IPv6 support disabled!" << std::endl;
      }
   }
   if(port < 1024) {
      std::cerr << "ERROR: Invalid port number!" << std::endl;
      exit(1);
   }
   if(timeout < 5000000) {
      timeout = 5000000;
   }
   else if(timeout > 1800000000) {
      timeout = 1800000000;
   }
   if(maxPacketSize < 256) {
      maxPacketSize = 256;
   }
   else if(maxPacketSize > 1024 * 1024) {
      maxPacketSize = 1024 * 1024;
   }


   // ====== Initialize QoS manager =========================================
   if(disableQM == false) {
      if(logName != NULL) {
         logStream = new std::ofstream(logName);
         if((logStream == NULL) || (!logStream->good())) {
            std::cerr << "ERROR: Unable to create log file!" << std::endl;
            exit(1);
         }
      }

      // ====== Check EUID ==================================================
      uid_t uid = geteuid();
      if(uid != 0) {
         std::cerr << "ERROR: Only root is allowed to run this program!" << std::endl;
         exit(1);
      }

      // ====== Open sockets ================================================
      pingSocket4 = new Socket(Socket::IPv4,Socket::Raw,Socket::ICMPv4);
      if(pingSocket4 == NULL) {
         std::cerr << "ERROR: Server::main() - Out of memory!" << std::endl;
         cleanUp(1);
      }
      if(!pingSocket4->bind()) {
         std::cerr << "ERROR: Unable to bind socket for ICMPv4!" << std::endl;
         exit(1);
      }
      pingSocket6 = NULL;
      if(InternetAddress::hasIPv6()) {
         pingSocket6 = new Socket(Socket::IPv6,Socket::Raw,Socket::ICMPv6);
         if(pingSocket6 == NULL) {
            std::cerr << "ERROR: Out of memory!" << std::endl;
            exit(1);
         }
         if(!pingSocket6->bind()) {
            std::cerr << "ERROR: Unable to bind socket for ICMPv6!" << std::endl;
            exit(1);
         }
      }

      // ====== Create RoundTripTimePinger =====================================
      pinger = new RoundTripTimePinger(pingSocket4,pingSocket6);
      if(pinger == NULL) {
         std::cerr << "ERROR: Server::main() - Out of memory!" << std::endl;
         cleanUp(1);
      }
      if(!pinger->ready()) {
         std::cerr << "ERROR: RoundTripTimePinger not ready!" << std::endl;
         exit(1);
      }

      // ====== Initialize QoS manager ======================================
      sla = new ServiceLevelAgreement();
      if(sla->load(slaFile.getData()) == false) {
         std::cerr << "ERROR: Unable to load SLA configuration file <"
                   << slaFile << ">!"<< std::endl;
         cleanUp(1);
      }
      qosManager = new BandwidthManager(sla,NULL);
      if(qosManager == NULL) {
         std::cerr << "ERROR: Server::main() - Out of memory!" << std::endl;
         cleanUp(1);
      }
      qosManager->setLogStream(logStream);
      qosManager->setFairness(fairnessSession,fairnessStream);
      qosManager->setQoSOptimizationParameters(maxRUPoints,utThreshold,bwThreshold,sdTolerance,unlayered);
      qosManager->setPartialRemapping(prEnabled,prReservedPortion,
                                     prUtilizationTolerance,prMaxRemappingInterval);
      qosManager->getFairness(fairnessSession,fairnessStream);
      qosManager->getQoSOptimizationParameters(maxRUPoints,utThreshold,bwThreshold,sdTolerance,unlayered);
      qosManager->getPartialRemapping(prEnabled,prReservedPortion,
                                     prUtilizationTolerance,prMaxRemappingInterval);
      qosManager->start();
#if 0
      pinger->start();
#endif
   }


   // ====== Initialize =====================================================
   initAll(directory.getData(), port,
           timeout, maxPacketSize, lossScalability,
           optUseSCTP);
#ifndef FAST_BREAK
   installBreakDetector();
#endif

   InternetAddress ourAddress;
   rtcpServerSocket->getSocketAddress(ourAddress);
   ourAddress.setPrintFormat(InternetAddress::PF_Address);


   // ====== Print status ===================================================
   std::cout << "RTP Audio Server - Copyright (C) 1999-2026 Thomas Dreibholz" << std::endl
             << "-----------------------------------------------------------" << std::endl
             << std::endl;
   if(optUseSCTP) {
      std::cout << "SCTP:             on" << std::endl;
   }
   else {
      std::cout << "SCTP:             off" << std::endl;
   }
   std::cout << "Server Port:      " << port << std::endl;
   char str[32];
   snprintf((char*)&str,sizeof(str),"$%08x",server->getOurSSRC());
   std::cout << "Server SSRC:      " << str << std::endl
             << "Client Timeout:   " << (timeout / 1000000) << " [s]" << std::endl
             << "Input Directory:  " << directory << std::endl
             << "Max Packet Size:  " << maxPacketSize << std::endl
             << "Loss Scalability: " << (lossScalability ? "on" : "off") << std::endl
             << std::endl;


   // ====== Main loop ======================================================
   for(;;) {
#ifndef FAST_BREAK
      if(breakDetected())
         break;
#endif
      Thread::delay(10000000,true);
   }


   // ====== Clean up =======================================================
   cleanUp(0);
}
